src/mono/mono/mini/interp/transform.c

```diff
@@ -10054,61 +10038,92 @@ add_active_call (TransformData *td, ActiveCalls *ac, InterpInst *call)
+ * where the base offset of a call must be greater than the offset of any argument of other active call args.
+ *
+ * Function first removes the call from an array of active calls. If a match is found,
+ * the call is removed from the array by moving the last entry into its place. Otherwise, it is a call without arguments.
+ *
+ * Call is push onto the stack with an array of other active calls that need to be resolved first.
+ *
+ * If there are no other active calls, function retrieves calls from the stack and resolves them, including the current call and all deferred calls.
+ *
+ * For each call, function computes the call offset of each call argument starting from a base offset, and stores the computed call offset into a InterpInst.
+ * The base offset is computed as max offset of all call offsets on which the call depends.
+ * Stack ensures that all call offsets on which the call depends are calculated before the call resolution.
+ */
 static void
 end_active_call (TransformData *td, ActiveCalls *ac, InterpInst *call)
 {
 	// Remove call from array
 	for (int i = 0; i < ac->active_calls_count; i++) {
-		if (ac->active_calls [i].call == call) {
+		if (ac->active_calls [i] == call) {
 			ac->active_calls_count--;
-			ac->param_size -= ac->active_calls [i].param_size;
 			// Since this entry is removed, move the last entry into it
 			if (ac->active_calls_count > 0 && i < ac->active_calls_count)
 				ac->active_calls [i] = ac->active_calls [ac->active_calls_count];
+			break;
 		}
 	}
-	// This is the relative offset (to the start of the call args stack) where the args
-	// for this call reside.
-	int start_offset = ac->param_size;
 
-	// Compute to offset of each call argument
-	int *call_args = call->info.call_args;
-	if (call_args && (*call_args != -1)) {
-		int var = *call_args;
-		while (var != -1) {
-			alloc_var_offset (td, var, &start_offset);
-			call_args++;
-			var = *call_args;
-		}
-	} else {
-		// This call has no argument. Allocate a dummy one so when we resolve the
-		// offset for MINT_CALL_ARGS_SREG during compacted instruction emit, we can
-		// always use the offset of the first var in the call_args array
-		int new_var = create_interp_local (td, mono_get_int_type ());
-		td->locals [new_var].call = call;
-		td->locals [new_var].flags |= INTERP_LOCAL_FLAG_CALL_ARGS;
-		alloc_var_offset (td, new_var, &start_offset);
-
-		call_args = (int*)mono_mempool_alloc (td->mempool, 3 * sizeof (int));
-		call_args [0] = new_var;
-		call_args [1] = -1;
+	// Push active call that should be resolved onto the stack
+	call->call_info = (InterpCallInfo*)mono_mempool_alloc (td->mempool, sizeof (InterpCallInfo));
+	call->call_info->call_deps = NULL;
+	for (int i = 0; i < ac->active_calls_count; i++)
+		call->call_info->call_deps = g_slist_prepend_mempool (td->mempool, call->call_info->call_deps, ac->active_calls [i]);
+	ac->deferred_calls = g_slist_prepend_mempool (td->mempool, ac->deferred_calls, call);
```

Comment:
We seem to be allocating a list entry for the most common case of no active calls / resolve arg offsets immediately.